library/bigip_software.py

```diff
@@ -212,673 +290,1060 @@
-
-        if state == 'activated' or state == 'installed':
-            if not volume:
-                if not reuse_inactive_volume:
-                    raise F5ModuleError("You must specify a volume")
-            elif not software:
-                raise F5ModuleError("You must specify a base image")
-
-        if state == "activated":
-            changed = self.activated()
-        elif state == "installed":
-            changed = self.installed()
-        elif state == "present":
-            changed = self.present()
-        elif state == "absent":
-            changed = self.absent()
-
-        if state in ['activated', 'installed', 'present']:
-            if not self.params['check_mode']:
-                current = self.read()
-                result.update(current)
-
-        changed = True
-        result.update(dict(changed=changed))
-        return result
+    def _find_iso_content(self, iso):
+        paths = ['/METADATA.XML', 'metadata.xml']
 
+        for path in paths:
+            try:
+                content = iso.record(path).content
+                return content
+            except KeyError:
+                pass
+        raise Exception(
+            "Unable to find metadata file in ISO. Please file a bug."
+        )
 
-class BigIpSoapApi(BigIpCommon):
-    def __init__(self, *args, **kwargs):
-        super(BigIpSoapApi, self).__init__(*args, **kwargs)
+    def handle_remote_source(self, link):
+        source = urlparse.urlparse(link)
+        name = source.path.split('/')[-1]
+        if source.scheme:
+            if source.scheme == 'http':
+                return name
+            elif source.scheme == 'https':
+                self._values['secure'] = True
+                return name
+            else:
+                err = 'Only remote HTTP or HTTPS sources are supported.'
+                raise F5ModuleError(err)
 
-        self.api = bigip_api(kwargs['server'],
-                             kwargs['user'],
-                             kwargs['password'],
-                             kwargs['validate_certs'])
+    def _list_volumes(self):
+        volumes = self.client.api.tm.sys.software.volumes.get_collection()
+        return volumes
 
-    def read(self):
-        result = dict(
-            software=[],
-            hotfix=[]
-        )
+    def _load_volume(self, volume):
+        vol = self.client.api.tm.sys.software.volumes.volume.load(
+            name=volume)
+        return vol
+
+    def _volume_exists(self, volume):
+        return self.client.api.tm.sys.software.volumes.volume.exists(
+            name=volume)
+
+    def _check_active_volume(self, volume):
+        v = self._load_volume(volume)
+        if hasattr(v, 'active') and v.active is True:
+            err = "Cannot install software or hotfixes to active volumes."
+            raise F5ModuleError(err)
+
+    def _get_next_available_volume(self):
+        volumes = self._list_volumes()
+        if len(volumes) == 1:
```

Comment:
no it is not redundant, if the volumes list is 1 then we perform volume name increment on Active volume only. If the list is >1 then we just return next non-active volume combining this will break this logic and mess it up